### PR TITLE
pluginupdate.py: add support for adding/updating individual plugins

### LINKED ISSUE #218584: improve vim plugin updater
## Issue description

I would like to be able to call the vim updater to update specific plugins and only these plugins (same for luarocks updater).
For instance in https://github.com/NixOS/nixpkgs/pull/218578, I would like to just update these plugins without regenerating the full set.

Which is anyway impossible at the moment because github flagged some github accounts as suspicious and made those repositories unavailable:
https://github.com/orgs/community/discussions/48173#discussioncomment-5117119
```
1183 plugins were checked, 3 plugin(s) could not be downloaded:

PluginDesc(repo=Repo(Spacegray.vim, https://github.com/ackyshake/Spacegray.vim/), branch='', alias=''): HTTP Error 404: Not Found
Traceback (most recent call last):
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 602, in prefetch
    plugin, redirect = prefetch_plugin(pluginDesc, cache)
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 477, in prefetch_plugin
    commit, date = repo.latest_commit()
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 71, in f_retry
    return f(*args, **kwargs)
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 187, in latest_commit
    with urllib.request.urlopen(commit_req, timeout=10) as req:
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
PluginDesc(repo=Repo(VimCompletesMe, https://github.com/ackyshake/VimCompletesMe/), branch='', alias=''): HTTP Error 404: Not Found
Traceback (most recent call last):
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 602, in prefetch
    plugin, redirect = prefetch_plugin(pluginDesc, cache)
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 477, in prefetch_plugin
    commit, date = repo.latest_commit()
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 71, in f_retry
    return f(*args, **kwargs)
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 187, in latest_commit
    with urllib.request.urlopen(commit_req, timeout=10) as req:
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
PluginDesc(repo=Repo(Vundle.vim, https://github.com/VundleVim/Vundle.vim/), branch='', alias=''): HTTP Error 404: Not Found
Traceback (most recent call last):
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 602, in prefetch
    plugin, redirect = prefetch_plugin(pluginDesc, cache)
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 477, in prefetch_plugin
    commit, date = repo.latest_commit()
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 71, in f_retry
    return f(*args, **kwargs)
  File "/home/teto/nixpkgs3/maintainers/scripts/pluginupdate.py", line 187, in latest_commit
    with urllib.request.urlopen(commit_req, timeout=10) as req:
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
```

On a sidenote, I wondered if we could add the script to the root's flake.nix's application, so that one can run `nix run .#update-vim-plugins`. This way we can enjoy nix autocompletion, it helps with discovery and one doesn't need to `fd update.py` to find the updater script 

### Steps to reproduce
`$ pkgs/applications/editors/vim/plugins/update.py --proc 5`

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

### LINKED ISSUE #218696: luarocks-package-updater improvements
## Issue description

Thanks to @mrcjkb 's work https://github.com/nvim-neorocks/luarocks-tag-release , we should (I hope !) see more neovim plugins on luarocks.org .

I have some issues with the luarocks updater, mainly it's slow.
One way around I would be happy with is if I could update (unpin)/add only one target, similiar to what I mention in  https://github.com/NixOS/nixpkgs/issues/218584

- [ ] one issue  is that if you run several luarocks process in parallel, they seem to overwrite their own manifest. Investigate/check htat's the case and see for a workaround (have each luarocks run with different  HOME or a proper upstream fix)
```
maintainers/scripts/update-luarocks-packages --github-token "$GITHUB_TOKEN" --proc 1
```

- [ ] the updater repeats this not nice warning:

```
Warning: Lua 5.1 interpreter not found at /nix/store/ffzdladqwg9xps985k7afy64w21skprr-lua-5.2.4

Modules may not install with the correct configurations. You may want to configure the path prefix to your build of Lua 5.1 using

   luarocks config --local lua_dir <your-lua-prefix>
```

- [ ] we should not create the derivation if it doesn't support that luaVersion (the disabled field); current implementation makes for confusing nixpkgs-review outputs:
```
7 packages marked as broken and skipped:
lua52Packages.telescope-manix lua52Packages.telescope-nvim lua53Packages.telescope-manix lua53Packages.telescope-nvim lua54Packages.haskell-tools-nvim lua54Packages.telescope-manix lua54Packages.telescope-nvim

8 packages built:
lua51Packages.haskell-tools-nvim lua51Packages.telescope-manix lua51Packages.telescope-nvim lua52Packages.haskell-tools-nvim lua53Packages.haskell-tools-nvim luajitPackages.haskell-tools-nvim luajitPackages.telescope-manix luajitPackages.telescope-nvim
```

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

### PR DESCRIPTION
**Pull request by PerchunPak**
## Description of changes

This PRs adds long-awaited support for updating individual plugins to Vim and Kakoune (`LuaEditor` is a third class that uses `pluginupdate.py`, but they patch too much in `pluginupdate`, so I leave it to someone else).

As a bonus (though it was my main goal), I implemented adding an entry to `generated.nix` when adding a plugin using the script. I also implemented a whole bunch of bugfixes and I will document some things as comments here.

Feel free to correct grammar, I am not a native, so will gladly accept any grammar nitpicks.

Fixes #218584.

## Things done

Things tested:

- Vim
  - [x] Full update ([result file](https://gist.github.com/PerchunPak/4771f491af3eaced66710c2306b06bb7))
  - [x] Individual update
  - [x] Adding package adds an entry to `generated.nix`
- Kakoune - seems broken on master, so leaving it on maintainer (@philiptaron)
- Lua
  - [ ] Full update - ratelimits even with token and one worker, assuming works

---

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC @teto (author of #218696), @figsoda (maintainer of Vim), @philiptaron (maintainer of Kakoune)

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: maintainers/scripts/pluginupdate.py [hunk 1]
```diff
@@ -425,27 +443,101 @@ def generate_nix(self, _plugins, _outfile: str):
...
+            plugin: Plugin on which you decide whether to ignore or not.
+            to_update:
+                List of strings, that user passed to `--update` parameter.
+                By default, we assume it is a list of URIs identical to what
+                is in input file.
+
+        Returns:
+            True if we should update plugin and False if not.
+        """
+        return plugin.name.replace(".", "-") in to_update
```
Comment (PerchunPak) [inline maintainers/scripts/pluginupdate.py:475]: Here you can see how optimistic I was for this PR (26 lines of docs to one line of code)

File: maintainers/scripts/pluginupdate.py [hunk 2]
```diff
@@ -559,6 +657,7 @@ def run(
         parser = self.create_parser()
         args = parser.parse_args()
         command = args.command or "update"
+        logging.basicConfig()
```
Comment (PerchunPak) [inline maintainers/scripts/pluginupdate.py:660]: Without this line, any logging using `log` variable didn't work. Also, why default log level is warning? It is not annoying on info level, should I change it too?

File: maintainers/scripts/pluginupdate.py [hunk 3]
```diff
@@ -382,16 +382,21 @@ def add(self, args):
         fetch_config = FetchConfig(args.proc, args.github_token)
         editor = self
         for plugin_line in args.add_plugins:
-            log.debug("using plugin_line", plugin_line)
+            log.debug("using plugin_line %s", plugin_line)
             pdesc = PluginDesc.load_from_string(fetch_config, plugin_line)
-            log.debug("loaded as pdesc", pdesc)
+            log.debug("loaded as pdesc %s", pdesc)
```
Comment (PerchunPak) [inline maintainers/scripts/pluginupdate.py:387]: It raised an error where complained at the lack of `%s`

File: maintainers/scripts/pluginupdate.py [hunk 4]
```diff
@@ -425,27 +443,101 @@ def generate_nix(self, _plugins, _outfile: str):
...
             return redirects
 
         return update
 
+    def merge_results(
+        self,
+        current: list[Tuple[PluginDesc, Plugin]],
+        fetched: List[Tuple[PluginDesc, Union[Exception, Plugin], Optional[Repo]]],
+    ) -> List[Tuple[PluginDesc, Union[Exception, Plugin], Optional[Repo]]]:
+        result: Dict[str, Tuple[PluginDesc, Union[Exception, Plugin], Optional[Repo]]] = dict(
```
Comment (PerchunPak) [inline maintainers/scripts/pluginupdate.py:530]: Those types are copied from `check_results` func, and result of this function will be passed to it

File: maintainers/scripts/pluginupdate.py [hunk 5]
```diff
@@ -753,15 +845,26 @@ def rewrite_input(
...
             if old_plugin.normalized_name != new_plugin.normalized_name:
                 deprecations[old_plugin.normalized_name] = {
                     "new": new_plugin.normalized_name,
                     "date": cur_date_iso,
                 }
+
+            # remove plugin from index file, so we won't add it
+            # to deprecations again
+            for i, plugin in enumerate(plugins):
+                if plugin.name == pdesc.name:
```
Comment (PerchunPak) [inline maintainers/scripts/pluginupdate.py:863]: As you could see at the description of [third commit](https://github.com/NixOS/nixpkgs/pull/336137/commits/9c05ff4ec3e4ec8e909786dba5af4a99d8f14bef), someone just forgot about deleting plugins from index for years!

File: maintainers/scripts/pluginupdate.py [hunk 6]
```diff
@@ -772,7 +875,7 @@ def rewrite_input(
         fieldnames = ["repo", "branch", "alias"]
         writer = csv.DictWriter(f, fieldnames, dialect="unix", quoting=csv.QUOTE_NONE)
         writer.writeheader()
-        for plugin in sorted(plugins):
+        for plugin in sorted(plugins, key=lambda x: x.name):
```
Comment (PerchunPak) [inline maintainers/scripts/pluginupdate.py:878]: interesting fact: plugin's URI is not unique, but name is. It is because there can be multiple plugins in one repo but on different branches (e.g. different versions, see [harpoon](https://github.com/NixOS/nixpkgs/blob/9caabb561519db70906c1435a3842417f519f1d9/pkgs/applications/editors/vim/plugins/vim-plugin-names#L386-L387)) and name is used as a key in the final result

File: pkgs/applications/editors/kakoune/plugins/update.py [hunk 7]
```diff
@@ -21,30 +21,40 @@
...
+lib.mapAttrs parse generated"""
 
 HEADER = "# This file has been generated by ./pkgs/applications/editors/kakoune/plugins/update.py. Do not edit!"
 
 class KakouneEditor(pluginupdate.Editor):
-
-
     def generate_nix(self, plugins: List[Tuple[pluginupdate.PluginDesc, pluginupdate.Plugin]], outfile: str):
-        sorted_plugins = sorted(plugins, key=lambda v: v[1].name.lower())
-
```
Comment (PerchunPak) [inline pkgs/applications/editors/kakoune/plugins/update.py:47]: I decided to move sorting to base class, because why duplicate logic? Furthermore, Vim and Kakoune sorted by different keys (also, Vim's sorting was broken)

File: pkgs/applications/editors/vim/plugins/update.py [hunk 8]
```diff
@@ -30,11 +30,6 @@
 
 log = logging.getLogger("vim-updater")
 
-sh = logging.StreamHandler()
-formatter = logging.Formatter("%(name)s:%(levelname)s: %(message)s")
-sh.setFormatter(formatter)
-log.addHandler(sh)
```
Comment (PerchunPak) [inline pkgs/applications/editors/vim/plugins/update.py:36]: This is unnecessary with `logging.basicConf` in the main file

**Comment by mweinelt** [comment]
@ofborg eval

File: maintainers/scripts/pluginupdate.py (hunk 2 above)
Comment (philiptaron) [reply maintainers/scripts/pluginupdate.py:660]: Logging at info level is a good idea.

File: maintainers/scripts/pluginupdate.py (hunk 5 above)
Comment (philiptaron) [reply maintainers/scripts/pluginupdate.py:863]: Time for some cleaning 🧹

File: maintainers/scripts/pluginupdate.py [hunk 9]
```diff
@@ -797,7 +900,12 @@ def update_plugins(editor: Editor, args):
         "Either set --proc=1 or --github-token=YOUR_TOKEN. ")
 
     fetch_config = FetchConfig(args.proc, args.github_token)
-    update = editor.get_update(args.input_file, args.outfile, fetch_config)
+    update = editor.get_update(
+        args.input_file,
+        args.outfile,
+        fetch_config,
+        getattr(args, "update_only", None),  # if script was called without arguments
```
Comment (philiptaron) [inline maintainers/scripts/pluginupdate.py:907]: I recommend using names for the parameters here.
```suggestion
        input_file=args.input_file,
        output_file=args.outfile,
        config=fetch_config,
        to_update=getattr(args, "update_only", None),  # if script was called without arguments
```

File: pkgs/applications/editors/kakoune/plugins/update.py [hunk 10]
```diff
@@ -21,30 +21,40 @@
 )
 import pluginupdate
 
-GET_PLUGINS = f"""(with import <localpkgs> {{}};
+GET_PLUGINS = f"""with import <localpkgs> {{ }};
```
Comment (philiptaron) [inline pkgs/applications/editors/kakoune/plugins/update.py:24]: This is the broken updater?

**Review by philiptaron — COMMENTED** [review]
Some initial feedback. This looks great -- thanks for your work here @PerchunPak.

What remains:

1. I need to validate each URL update
2. You mentioned that I have some work to do for kakoune. Could you make it more clear for me what's broken and what I ought to pay attention to?
3. This deprecated.json thing is a total mess. What would cleanup look like?

Thank you!

File: pkgs/applications/editors/kakoune/plugins/update.py (hunk 10 above)
Comment (PerchunPak) [reply pkgs/applications/editors/kakoune/plugins/update.py:24]: Yes, if you run

```bash
cd pkgs/applications/editors/kakoune/plugins
nix-shell update-shell.nix
python update.py
```

It outputs (on master and here is the same error)
```nix
error:
       … <borked>

         at «none»:0: (source not available)

       … from call site

         at «string»:1:6:

            1| with import <localpkgs> { };
             |      ^
            2| let

       error: function 'anonymous lambda' called without required argument 'kakouneUtils'

       at /home/perchun/dev/nixpkgs/pkgs/applications/editors/kakoune/plugins/default.nix:1:1:

            1| { callPackage, config, kakouneUtils, lib }:
             | ^
            2|
```

File: maintainers/scripts/pluginupdate.py [hunk 11]
```diff
@@ -425,27 +443,101 @@ def generate_nix(self, _plugins, _outfile: str):
...
+        return plugin.name.replace(".", "-") in to_update
+
+    def get_update(self, input_file: str, outfile: str, config: FetchConfig, to_update: Optional[List[str]]):
+        if to_update is None:
+            to_update = []
+
+        current_plugins = self.get_current_plugins(config, self.nixpkgs)
+        current_plugin_specs = self.load_plugin_spec(config, input_file)
+
+        cache: Cache = Cache(map(lambda x: x[1], current_plugins), self.cache_file)
```
[the remark this answers is not in the document]
Comment (PerchunPak) [reply maintainers/scripts/pluginupdate.py:484]: Totally forgot that there is a more pythonic (and faster! because creating function is not cheap) way to filter/map

**Comment by PerchunPak** [comment]
Pushed [changes from review](https://github.com/NixOS/nixpkgs/compare/65d9fa3f7237998ae46320781a4d450fbb6b3729..03d32456ebf113809645c8747e9b07230fdd46ff) and [formatting changes](https://github.com/NixOS/nixpkgs/compare/03d32456ebf113809645c8747e9b07230fdd46ff..16419baded8cb4375f51b737cf8650575c2cc73f) in different force pushes for simpler reviewing. I didn't push it as a new commit, as it would create a lot of conflicts if there will be other changes.

**Comment by PerchunPak** [comment]
> I need to validate each URL update

This should be easy with `nixpkgs-review`:

Result of `nixpkgs-review pr 336137` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>luarocks-packages-updater</li>
    <li>nixpkgs-manual</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

Which means no Vim plugins were changed/removed/added. I also quickly glanced over the diff on generated.nix after the full plugin update and didn't find any plugins that were moved/removed/added (before I created the PR there were a few bugs with sorting).

> This deprecated.json thing is a total mess. What would cleanup look like?

What do you mean? The file `deprecated.json` or the code for it? If the code, it is in a good state compared to other parts of the code I've seen in this file.

File: maintainers/scripts/pluginupdate.py [hunk 12]
```diff
@@ -425,27 +443,101 @@ def generate_nix(self, _plugins, _outfile: str):
         """Returns nothing for now, writes directly to outfile"""
         raise NotImplementedError()
 
-    def get_update(self, input_file: str, outfile: str, config: FetchConfig):
-        cache: Cache = Cache(self.get_current_plugins(self.nixpkgs), self.cache_file)
+    def filter_plugins_to_update(self, plugin: PluginDesc, to_update: List[str]) -> bool:
+        """Function for filtering out plugins, that user doesn't want to update.
```
[the remark this answers is not in the document]
Comment (PerchunPak) [reply maintainers/scripts/pluginupdate.py:458]: I am pretty sure this breaks the [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) I used in this docstring. Though the closest rule I could find is https://docs.astral.sh/ruff/rules/no-signature/

File: maintainers/scripts/pluginupdate.py [hunk 13]
```diff
@@ -328,12 +328,11 @@ def load_plugins_from_csv(
...
 
 
-
 def run_nix_expr(expr, nixpkgs: str, **args):
-    '''
+    """
     :param expr nix expression to fetch current plugins
     :param nixpkgs Path towards a nixpkgs checkout
-    '''
+    """
```
Comment (teto) [inline maintainers/scripts/pluginupdate.py:335]: aws this changed by black ?

File: doc/languages-frameworks/vim.section.md [hunk 14]
```diff
@@ -251,10 +251,15 @@ nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --github-token=mytoken
...
 
 ```sh
-
 nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --proc 1'
 ```
 
+If you want to update only certain plugins, you can specify their names to `--update` flag. Note that those name must be the same as in `pkgs/applications/editors/vim/plugins/vim-plugin-names` file.
+
+```sh
+nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --update "nvim-treesitter" "LazyVim"'
```
Comment (teto) [inline doc/languages-frameworks/vim.section.md:260]: apparently it's not --update but update:
```
nix run .#vimPluginsUpdater --  update nvim-treesitter LazyVim
WARNING:root:You have enabled parallel updates but haven't set a github token.
You may be hit with `HTTP Error 429: too many requests` as a consequence.Either set --proc=1 or --github-token=YOUR_TOKEN. 
path is '/nix/store/rgwip825v0m4nhsvqyabf0mb2as81x0q-0b8b78f9d08dc338a146eb4cd4bcbed8dd36a783.tar.gz'
2 of 1471 were checked
```

**Comment by teto** [comment]
thanks for the great work. It's a very exciting feature. I had thought of a treesitter-based solution but this seems to work as well. Not having the feature in all updaters is no problem, that's part of the design. 

I am a bit annoyed that you have mixed different things in the PR (logging level, formatting, sorting, new feature). Most of it I want to merge but some of it (the update filtering) I would like to test & experiment with myself more. Would it be too much to ask for the sorting fix in a separate PR ? 
For the logging part, the approach is to `print` user-facing messages and logging is a separate thing, which is why the level is set to WARN (since the logger is verbose, which is ok, it's targeted at developers first).
As for the formatting, it's supposed to be "black"-ified but maybe some parts were not done correctly. 
If you could open separate PRs for the different changes, it would help with review work.

**Comment by PerchunPak** [comment]
> I am a bit annoyed that you have mixed different things in the PR (logging level, formatting, sorting, new feature). ... Would it be too much to ask for the sorting fix in a separate PR ?

Unfortunately, most of those changes are QOL and usually they fix some weird bug, that prevents from properly implementing individual updates. For example, before sorting was completely broken (it sorted by `PluginDesc.name`, `PluginDesc.name` implementation was:
```py
if self.alias is not None:
    return self.alias
else:
    return self.repo.name
```
BUT `self.alias`, if unset, was an empty string (because in CSV if something is unset, it is an empty string), instead of `None`, so it sorted by what request finished first, instead of proper name). This resulted in `vim-updater-script add ...` resorts half a file, if ran right after `vim-updater-script update`. Implementing proper sorting prevents merge conflicts. And I don't see much sense in individual updates, if those generate huge diffs that are hard to review.

> For the logging part, the approach is to print user-facing messages and logging is a separate thing, which is why the level is set to WARN (since the logger is verbose, which is ok, it's targeted at developers first).

Ok, I will revert `INFO` by default.

File: maintainers/scripts/pluginupdate.py (hunk 13 above)
Comment (PerchunPak) [reply maintainers/scripts/pluginupdate.py:335]: I ran `ruff format` which is [99.9%](https://docs.astral.sh/ruff/faq/#how-does-ruffs-formatter-compare-to-black) compatible with black (just because it was installed on my system), so whatever?

File: doc/languages-frameworks/vim.section.md (hunk 14 above)
Comment (PerchunPak) [reply doc/languages-frameworks/vim.section.md]: Fixed [here](https://github.com/NixOS/nixpkgs/compare/0c322416d8569700c1a24589e060a04243eb581c..b306fc0cf6fc572163b1da98834def72621c9432), thanks!

**Comment by PerchunPak** [comment]
I am reviewing (and merging with the help of GaetanLepage) all `vimPlugins.*` PRs, so conflicts here are probably to stay until merge (they are not worth solving, if next merged PR will break them again + I will need to ensure sorting is correct right before merge)

File: pkgs/applications/editors/kakoune/plugins/update.py (hunk 7 above)
Comment (teto) [reply pkgs/applications/editors/kakoune/plugins/update.py:47]: agreed

File: pkgs/applications/editors/kakoune/plugins/update.py (hunk 10 above)
Comment (philiptaron) [reply pkgs/applications/editors/kakoune/plugins/update.py:24]: Thanks! It's on the list to go fix. I'm camping this weekend but will attend to it when back.

**Comment by PerchunPak** [comment]
Fixed merge conflicts and added support for [this](https://github.com/NixOS/nixpkgs/pull/344478#discussion_r1786118486)

File: doc/languages-frameworks/vim.section.md [hunk 15]
```diff
@@ -251,10 +251,15 @@ nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --github-token=mytoken
 Alternatively, set the number of processes to a lower count to avoid rate-limiting.
 
 ```sh
-
 nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --proc 1'
 ```
 
+If you want to update only certain plugins, you can specify their names to `update` flag. Note that those name must be the same as in `pkgs/applications/editors/vim/plugins/vim-plugin-names` file.
```
Comment (MattSturgeon) [inline doc/languages-frameworks/vim.section.md:257]:
```suggestion
If you want to update only certain plugins, you can specify them after the `update` command. Note that you must use the same plugin names as the `pkgs/applications/editors/vim/plugins/vim-plugin-names` file.
```

**Review by MattSturgeon — APPROVED** [review]
I'm not too familiar, but I don't see any major issues!

As requested, I've left a suggestion for documentation wording. Take as much or as little from that as you please :slightly_smiling_face:

File: doc/languages-frameworks/vim.section.md [hunk 16]
```diff
@@ -251,10 +251,15 @@ nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --github-token=mytoken
 Alternatively, set the number of processes to a lower count to avoid rate-limiting.
 
 ```sh
-
 nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --proc 1'
 ```
 
+If you want to update only certain plugins, you can specify them after the `update` command. Note that you must use the same plugin names as the `pkgs/applications/editors/vim/plugins/vim-plugin-names` file.
```
Comment (teto) [inline doc/languages-frameworks/vim.section.md:257]: we can remove this since single plugin update will be in a separate PR ?

Comment (PerchunPak) [reply doc/languages-frameworks/vim.section.md:257]: Done in #353786!

File: maintainers/scripts/pluginupdate-py/pluginupdate.py [hunk 17]
```diff
@@ -421,28 +453,117 @@ def generate_nix(self, _plugins, _outfile: str):
...
         return update
 
+    def merge_results(
+        self,
+        current: list[Tuple[PluginDesc, Plugin]],
+        fetched: List[Tuple[PluginDesc, Union[Exception, Plugin], Optional[Repo]]],
+    ) -> List[Tuple[PluginDesc, Union[Exception, Plugin], Optional[Repo]]]:
+        # transforming this to dict, so lookup is O(1) instead of O(n) (n is len(current))
+        result: Dict[
+            str, Tuple[PluginDesc, Union[Exception, Plugin], Optional[Repo]]
```
Comment (GaetanLepage) [inline maintainers/scripts/pluginupdate-py/pluginupdate.py:555]:
```suggestion
    ) -> list[tuple[PluginDesc, Union[Exception, Plugin], Repo | None]]:
        # transforming this to dict, so lookup is O(1) instead of O(n) (n is len(current))
        result: dict[
            str, tuple[PluginDesc, Union[Exception, Plugin], Repo | None]
```

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

File: maintainers/scripts/pluginupdate-py/pluginupdate.py (hunk 17 above)
Comment (PerchunPak) [reply maintainers/scripts/pluginupdate-py/pluginupdate.py]: Done in 3c2cc30, thanks!

File: doc/languages-frameworks/vim.section.md [hunk 18]
```diff
@@ -264,10 +264,15 @@ nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater --github-token=mytoken
...
+
+```sh
+nix-shell -p vimPluginsUpdater --run 'vim-plugins-updater update "nvim-treesitter" "LazyVim"'
```
[the remark this answers is not in the document]
Comment (PerchunPak) [reply doc/languages-frameworks/vim.section.md:273]: No, but it is easier to read if they are there. I can remove it if you want

[the remark this answers is not in the document]
Comment (GaetanLepage) [reply doc/languages-frameworks/vim.section.md:273]: I don't have a strong feeling, we can keep them.

**Review by GaetanLepage — APPROVED** [review]
The implementation looks good overall. Have you been able to successfully test it ?
